### ISSUE TITLE
Context Pagination

### ISSUE THREAD

**Issue by lambadalambda** [tootsuite/mastodon#11029]
### Pitch

Contexts should have some form of pagination. Anytime descendants or ancestors get over ~ 20, there should be link headers set to fetch the rest.

### Motivation

Some threads can get rather long and can even make the request time out if long enough. Even if they don't, waiting 5 seconds for a thread to load seems like a bad idea. As this would be a breaking change for contexts (suddenly clients will get less information returned) I'd suggest either doing that under the /v2 endpoints or adding a parameter `?paginated=true` to enable this paginated mode.

**Comment by lambadalambda** [comment]
Sister issue: https://git.pleroma.social/pleroma/pleroma/issues/969

**Comment by nightpool** [comment]
see also https://github.com/tootsuite/mastodon/pull/7564

we already have this feature, it's just currently set very high since we
don't have UI for it and don't want to break backwards compatibility

**Comment by lambadalambda** [comment]
So does this actually do pagination or just a cut-off?

**Comment by ThibG** [comment]
It allows cutting off and offers ways to paginate, but those ways are broken.

**Comment by nightpool** [comment]
there are a couple ways to paginate currently, but it looks like aki didn't
put any of them into the API, he moved on to other things before starting
the UI implementation, so we had to revert the limits.

I think I broadly agree with thibg's comments on GitHub that the applying
concept of pagination for a "flattened tree" like the one mastodon
implements for replies isn't particularly well-defined. the best user
experience would probably be limiting on thread depth from root, but then
we'd have to include multiple "expand this thread" buttons and an API to
specify where those buttons are.

**Comment by nightpool** [comment]
there are a couple ways to paginate currently, but it looks like aki didn't
put any of them into the API, he moved on to other things before starting
the UI implementation, so we had to revert the limits.


I think I broadly agree with thibg's comments on GitHub that the applying
concept of pagination for a "flattened tree" like the one mastodon
implements for replies isn't particularly well-defined. the best user
experience would probably be limiting on thread depth from root, but then
we'd have to include multiple "expand this thread" buttons and an API to
specify where those buttons are.

**Comment by lambadalambda** [comment]
Does anything actually depend on this tree nature? As far as I can tell, threads are displayed linearly everywhere. What you describe as the best user experience (I'd rather say 'most correct' way) seems very complicated to me (from an UI/UX standpoint.

Currently, Mastodon does a depth-first flattening of the downward tree, correct?

**Comment by nightpool** [comment]
yeah, that's correct, with chronological subtree ordering. the problem (aside from the UX concerns) is that I'm not even sure how we would implement the recursive query that displays replies as a resumable operation—a simple flat pagination

**Comment by nightpool** [comment]
I guess to be fully correct we actually need pagination in two directions—depth and width. I'm not convinced width is useful for the majority of use-cases but we should probably at least consider it

**Comment by lambadalambda** [comment]
I mean, this is mostly for the extreme threads anyway. Loading 100 or 200 post threads is fine, but anything approaching 1000 becomes unusable. If the 'paginated threads' endpoint would return the tree flattened by post id / post date order, pagination would be rather straight forward. Frontends can still reconstruct the old sorting for everything that fits into the pagination limit.

**Comment by nightpool** [comment]
sure, but then as you paginate new statuses "pop in" elsewhere in the tree. that's really confusing from a user perspective—especially for threads that are as long as you say they are. 

plus the invisible truncation would make it harder then it even currently is to understand which replies are part of which sub-threads.

**Comment by lambadalambda** [comment]
We have reply markers in pleroma, so this not really a problem for us. Still, would be nice to find a solution that works for everyone.

**Comment by nightpool** [comment]
I'm not sure how that's relevant, since you don't use the Mastodon API to
render the pleroma fe

**Comment by lambadalambda** [comment]
we do tho :)

**Comment by nightpool** [comment]
so do you have a different ordering of descendents or do you change the ordering client side?

**Comment by lambadalambda** [comment]
we just have it like i described above, by post id. I wasn't even aware that Mastodon did a more complicated ordering.

**Comment by antonpodolsky** [comment]
I'm starting to see statuses for which the replies take an extremely long time to load.

I took 2 minutes to fetch 160 replies in [this ](https://mastodon.social/@FranckLeroy@mastodon.online/109627512861474935) case:

![image](https://github.com/mastodon/mastodon/assets/7899596/e78d2c10-f2ce-4112-aa22-ac5e44266e2e)

**Comment by sneakers-the-rat** [comment]
This is a more humongous problem than it appears I think, it both makes threads very slow/inefficient on local instances (as noted above), but it also poses strong limits on things like backfilling replies (since it makes sense to not send 100% of everything to logged out users for DoS reasons). I'd argue that it is one of the highest ratios of how much it limits the fedi vs. how complicated it would be to solve.

There are a few ways forward and i want to pitch in this, so trying to reboot this conversation:
- serve posts chronologically - we would assume that the web ui and apps would be requesting most of these anyway, but still be able to take advantage of better caching and an unambiguous api. downsides are that this might make deep threads odd where posts pop in above where we are in the scroll view (again this noted above).
- serve posts in the tree order with an opaque pagination cursor - generate an opaque cursor that refers to the next/previous pages for a given thread at a particular point in time. then if there are new posts that should have appeared earlier in the thread sort those can be appended to the next page. this would be a trickier backend implementation, and would make incremental updates harder, but would resolve the ordering problems
- limit depth relatively low (like 3), show placeholder "expand thread" buttons, and serve chronologically - this reduces the deep thread ordering problem in a sort of "cheap" way with the added bonus of punting on the similarly dire need to rework the thread view in the web ui.
- split up the process - the `context` endpoint just serves URIs/IDs (and can thus serve a lot of them, potentially the whole tree) cheaply, and then we have a batched `statuses` endpoint that retrieves the full status for a batch of statuses. This is how [bsky is doing it](https://docs.bsky.app/docs/tutorials/viewing-threads) and it seems to work well. If we served just the thread skeleton (for each post, its uri, the uri of its parent, and the uris of its children) then we could *not* do any tree sorting in the `context` endpoint and just serve that chronologically, and then the web API could be in control of the tree sorting and put together status requests in batches on scroll. Along with a streaming endpoint for threads, then that would be pretty complete and fast and good i think. This also makes the activitypub collections representation more efficient and allows us to unify the behavior with the rest API, where instead of inlining collections of replies we just serve their URIs in the same way.

not all of these are mutually exclusive, but since this thread has mostly lagged since 2019 i wanted to get it rolling again and also see where we're at on this generally. I see labels being changed around within the last year or so, so maybe now is the time?

**Comment by rauschma** [comment]
@sneakers-the-rat If I think about scaling this up to a large forest of replies, my favorite user interface would be:
* Initially, only show the top level of the forest. This makes navigation more manageable if there are many reply threads. Subsequent levels can be expanded on demand.
* At all levels, nodes should probably be paged, with an option to jump to the end.
* Below the top level, branches where posts only have single replies, should be flattened. That helps with longer conversations, where it’s cumbersome if you have to expand each member of a chain of posts.

**Comment by sneakers-the-rat** [comment]
Agreed, except I think that typically having a depth of 2-3 rather than 1 at the top level is good to see if one wants to expand context, but that could be done with a cheap "skeleton" context that just returns status URIs, timestamps, and `in_reply_to` rather than the full status for the whole/a large chunk of the tree, and then allowing the ui to do batched requests for the status content afterwards. It's gotta be cheaper to make multiple requests and paginate than serve literally 4096 replies in one go, or whatever that limit is.

I think all of this would require masto to have a proper threading UI to make work in a satisfying way, but the context endpoint can still be repaired without it. I think mixing the UI concerns and backend concerns is sorta what got us here, where the backend does tree sort for display, which makes pagination hard, but there is no reason for backend to do tree sort without pagination because we always return the full thing and [sort it again anyway](https://github.com/mastodon/mastodon/blob/295ad6f19a016b3f16e1201ffcbb1b3ad6b455a2/app/javascript/mastodon/reducers/contexts.js#L18). So any strategy that gets backend out of the business of sorting I think would be an improvement
